src/discussions/topics/TopicsView.test.jsx

```diff
@@ -0,0 +1,186 @@
+      </AppProvider>
+    </IntlProvider>,
+  );
+}
+
+describe('TopicsView', () => {
+  beforeEach(() => {
+    initializeMockApp({
+      authenticatedUser: {
+        userId: 3,
+        username: 'abc123',
+        administrator: true,
+        roles: [],
+      },
+    });
+
+    store = initializeStore();
+    Factory.resetAll();
+    axiosMock = new MockAdapter(getAuthenticatedHttpClient());
+
+    lastLocation = undefined;
+  });
+
+  it('displays non-courseware topics', async () => {
+    axiosMock
+      .onGet(topicsApiUrl)
+      .reply(200, {
+        courseware_topics: [],
+        non_courseware_topics: Factory.buildList('topic', 3),
+      });
+    renderComponent();
+
+    await screen.findByText('topic 1');
+    expect(screen.queryByText('topic 2')).toBeInTheDocument();
+    expect(screen.queryByText('topic 3')).toBeInTheDocument();
+  });
+
+  it('displays non-courseware in one topic group', async () => {
+    axiosMock
+      .onGet(topicsApiUrl)
+      .reply(200, {
+        courseware_topics: [],
+        non_courseware_topics: Factory.buildList('topic', 3),
+      });
+    renderComponent();
+
+    await screen.findByText('topic 1');
+    const topicGroups = screen.queryAllByTestId('topic-group');
+    expect(topicGroups).toHaveLength(1);
+  });
+
+  it('displays courseware topics', async () => {
+    axiosMock
+      .onGet(topicsApiUrl)
+      .reply(200, {
+        courseware_topics: Factory.buildList('topic', 3),
+        non_courseware_topics: [],
+      });
+    renderComponent();
+
+    await screen.findByText('topic 1');
+    expect(screen.queryByText('topic 2')).toBeInTheDocument();
+    expect(screen.queryByText('topic 3')).toBeInTheDocument();
+  });
+
+  it('displays courseware topics in individual topic groups', async () => {
+    axiosMock
+      .onGet(topicsApiUrl)
+      .reply(200, {
+        courseware_topics: Factory.buildList('topic', 3),
+        non_courseware_topics: [],
+      });
+    renderComponent();
+
+    await screen.findByText('topic 1');
+    const topicGroups = screen.queryAllByTestId('topic-group');
+    expect(topicGroups).toHaveLength(3);
+  });
+
+  it('displays non-courseware topics before courseware topics', async () => {
+    axiosMock
+      .onGet(topicsApiUrl)
+      .reply(200, {
+        courseware_topics: Factory.buildList('topic', 3),
+        non_courseware_topics: Factory.buildList('topic', 3),
+      });
+    renderComponent();
+
+    await screen.findByText('topic 1');
+    const topicGroups = screen.queryAllByTestId('topic-group');
+    expect(queryByText(topicGroups[0], 'topic 4')).toBeInTheDocument();
+    expect(queryByText(topicGroups[0], 'topic 5')).toBeInTheDocument();
+    expect(queryByText(topicGroups[0], 'topic 6')).toBeInTheDocument();
+    expect(queryByText(topicGroups[1], 'topic 1')).toBeInTheDocument();
+    expect(queryByText(topicGroups[2], 'topic 2')).toBeInTheDocument();
+    expect(queryByText(topicGroups[3], 'topic 3')).toBeInTheDocument();
```

Comment:
You can directly access the element using text. so why you are using the `queryAllByTestId `? I'm confused why the courseware topics are on the 0 index and others are on the different indexes?

Reply:
Short answer is - non-courseware topics are all grouped under one 'topic-group', and courseware topics are treated like individual topic groups. Thus, the reason I query using test id, is to be able to check that:
1. The non-courseware topics are under separate topic group, which, I can agree, might be redundant since I'm already testing that in another test above
2. To be able to check the order of the topics displayed. I can't check the order (afaik, may be you can point me to how to do it) by query by text. And since in the DOM the non-courseware and courseware topics are represented differently, I can't just slap a test id on one element and check the order of that. I have to check by looking through the containers.
Let me know if it doesn't answer your questions, I will try to explain with examples.

Reply:
@Cup0fCoffee thanks for the explanation. I got your point. I can understand the position issue here.